17492: Increase SideNavMenu unit test coverage

mikdhadc opened the pull request:
Added testcases for SideNavMenu Component that increased the test coverage

BREAKING CHANGE: The SideNavMenu now has a test coverage greater than 90%

Closes #17492 


#### Changelog

**Changed**

- Added tests case to properly test and increase the test coverage for SideNavMenu

#### Testing / Reviewing
<img width="1776" alt="Screenshot 2024-10-24 at 7 10 14 PM" src="https://github.com/user-attachments/assets/d4e91b2d-3a41-4663-ad1e-01870cba808b">

mikdhadc committed:
test: added testcases for SideNavMenu

Added testcases for SideNavMenu Component that increased the test coverage

BREAKING CHANGE: The SideNavMenu now has a test coverage greater than 80%

mikdhadc committed:
Merge branch 'main' into 17492_SideNavMenu-unit-test

tay1orjones commented on packages/react/src/components/UIShell/__tests__/SideNavMenu-test.js:
```diff
@@ -98,4 +102,118 @@ describe('SideNavMenu', () => {
+      </SideNavMenuItem>,
+    ];
+    expect(hasActiveDescendant(child)).toBe(true);
+  });
```
Rather than test the internal implementation like this - how could we reframe the tests from the perspective of someone using the component?

hasActiveDescendant is only called in one place under certain circumstances
https://github.com/carbon-design-system/carbon/blob/061970723f64d1376ea28109328c1aec1f192ecc/packages/react/src/components/UIShell/SideNavMenu.tsx#L91-L92

I think this would mean that to test this from a user-interaction viewpoint, each test could `render` the component with a different type of children which would get `hasActiveDescendant` to be called, and then the assertion would be `toHaveClass('${prefix}--side-nav__item--active')`.

tay1orjones commented on packages/react/src/components/UIShell/__tests__/SideNavMenu-test.js:
```diff
@@ -98,4 +102,118 @@ describe('SideNavMenu', () => {
+  it('should return true if the child is a invalid React element and has isActive props set to true', () => {
+    const child = <div isActive={true} />;
+    expect(hasActiveDescendant(child)).toBe(true);
+  });
```
Same kind of idea here. I don't think we need to directly assert that the function call returns true. Instead it's the end result of the function call we want to ensure happens correctly.


Again I think in nearly all cases this would be an assertion of `toHaveClass` or `.not.toHaveClass`. The classes are admittedly still an implementation detail, but they're the closest thing to what the user perceives without having to actually assert that certain styles are applied (background, etc. or whatever is in selectors with `${prefix}--side-nav__item--active`

tay1orjones commented on packages/react/src/components/UIShell/__tests__/SideNavMenu-test.js:
```diff
@@ -98,4 +102,118 @@ describe('SideNavMenu', () => {
+      </SideNavMenu>
+    );
+    expect(hasActiveDescendant(child)).toBe(false);
+  });
```
This one doesn't use anything that's being rendered.

tay1orjones commented on packages/react/src/components/UIShell/SideNavMenu.tsx:
```diff
@@ -200,7 +200,9 @@ SideNavMenu.propTypes = {
 a single element, an array of elements, or null or undefined.
 **/
-function hasActiveDescendant(children: ReactNode | ReactNode[]): boolean {
+export function hasActiveDescendant(
```
If the tests are refactored to be [user centered](https://testing-library.com/docs/#:~:text=What%20you%20should,Child%20components) this won't need to be exported.

tay1orjones commented on packages/react/src/components/UIShell/__tests__/SideNavMenu-test.js:
```diff
@@ -98,4 +102,118 @@ describe('SideNavMenu', () => {
+      </SideNavContext.Provider>
+    );
+
+    fireEvent.keyDown(screen.getByText(/a/i), {
```
Could/should this use `userEvent` instead?

tay1orjones reviewed (changes requested):
Thanks for taking this on! I have a couple points of feedback below, let me know what you think.

mikdhadc committed:
Merge branch 'main' into 17492_SideNavMenu-unit-test

mikdhadc committed:
test: added testcases for SideNavMenu

Added testcases for SideNavMenu Component that increased the test coverage

BREAKING CHANGE: The SideNavMenu now has a test coverage greater than 90% - fixes based on feedback

mikdhadc committed:
fix: remove unused imports

Removed unused import fireEvent from the test file

BREAKING CHANGE: Removed unused imports

mikdhadc committed:
Merge branch 'main' into 17492_SideNavMenu-unit-test

tay1orjones commented on packages/react/src/components/UIShell/__tests__/SideNavMenu-test.js:
```diff
@@ -98,4 +101,116 @@ describe('SideNavMenu', () => {
+        <SideNavMenu
+          isSideNavExpanded={true}
+          defaultExpanded={true}
+          ref={mockRef} //uses the value of mockRef from previous test
```
Could you share more on why this is needed? Generally tests should be isolated from one another and if you need to mock a ref you can do it with jest.fn() like in the `should support a `ref` on the <button> element` test.

mikdhadc replied:
Thanks. It was unnecessary there. I have removed them.

mikdhadc committed:
fix: isolated tests

Two tests were using the same 'ref', which is removed to isolate the tests

BREAKING CHANGE: isolated tests

mikdhadc committed:
fix: remove unnecessary 'ref'

Removed ref from tests which are not necessary

BREAKING CHANGE: removed unnecessary ref

mikdhadc committed:
Merge branch 'main' into 17492_SideNavMenu-unit-test

mikdhadc committed:
Merge branch 'main' into 17492_SideNavMenu-unit-test

mikdhadc committed:
Merge branch 'main' into 17492_SideNavMenu-unit-test

mikdhadc committed:
Merge branch 'main' into 17492_SideNavMenu-unit-test

tay1orjones commented on packages/react/src/components/UIShell/__tests__/SideNavMenu-test.js:
```diff
@@ -98,4 +99,111 @@ describe('SideNavMenu', () => {
+      'aria-expanded',
+      'false'
+    );
+  });
```
I won't hold up the merging of this as the tests pass and do cover the code properly, but the naming and structure of these was really confusing to me. I had to open up the code coverage report to discern what you're aiming to cover. I found this:

![2024-10-29 at 13 36 55-MAIN-Google Chrome](https://github.com/user-attachments/assets/0ff409da-5e44-4ed2-b2cb-ca3756a9a538)

Refactoring the structure to be something like below would help

```js

describe('properly detects active descendants and applies active styling', () => {
  it('when a child in children array has isActive', () => { ... });
  it('when a child in children array has aria-current', () => { ... });
  it('when a child in nested children array has isActive', () => { ... });
  it('when a child in nested children array has aria-current', () => { ... });
  it('when a single child has isActive', () => { ... });
  it('when a single child has aria-current', () => { ... });
});

describe('properly detects non-active descendants and does not apply active styling', () => {
  it('when a child in children array is not a valid element', () => { ... });
  it('when all children have no props given', () => { ... });
  it('when a single child has no props', () => { ... });
});

```

Each of these would need a render block with `SideNavMenu defaultExpanded={true}`. Each assertion would still be `toHaveClass` of the active class, even when given `aria-current`. For the "children array" ones, pass in multiple SideNavMenuItem components rather than a dummy array.

mikdhadc replied:
Thanks for the feedback. I have updated the code. Please let me know if anything else is to be done. thank you
@tay1orjones @emyarod

mikdhadc committed:
Merge branch 'main' into 17492_SideNavMenu-unit-test

mikdhadc committed:
test: added testcases for SideNavMenu

Added testcases for SideNavMenu Component that increased the test coverage

BREAKING CHANGE: The SideNavMenu now has a test coverage of 100% (branch 92%)

emyarod reviewed (approved):
looks good to me!

mikdhadc committed:
Merge branch 'main' into 17492_SideNavMenu-unit-test

mikdhadc committed:
Merge branch 'main' into 17492_SideNavMenu-unit-test

tay1orjones committed:
Merge branch 'main' into 17492_SideNavMenu-unit-test
